src/exception.c

```diff
@@ -281,16 +289,97 @@ void __onCriticalException(volatile reg_block_t* regs)
 }
 
 /**
- * @brief Respond to a reset exception
+ * @brief Register a handler that will be called when the user
+ *        presses the RESET button. 
+ * 
+ * The N64 sends an interrupt when the RESET button is pressed,
+ * and then actually resets the console after about ~500ms (but less
+ * on some models, see #RESET_TIME_LENGTH).
+ * 
+ * Registering a handler can be used to perform a clean reset.
+ * Technically, at the hardware level, it is important that the RCP
+ * is completely idle when the reset happens, or it might freeze
+ * and require a power-cycle to unfreeze. This means that any
+ * I/O, audio, video activity must cease before #RESET_TIME_LENGTH
+ * has elapsed.
+ * 
+ * This entry point can be used by the game code to basically
+ * halts itself and stops issuing commands. Libdragon itself will
+ * register handlers to halt internal modules so to provide a basic
+ * good reset experience.
+ * 
+ * Handlers can use #exception_reset_time to read how much has passed
+ * since the RESET button was pressed.
+ * 
+ * @param cb    Callback to invoke when the reset button is pressed.
+ * 
+ * @note  Reset handlers are called under interrupt.
+ * 
  */
-void __onResetException(volatile reg_block_t* regs)
+void register_reset_handler( void (*cb)(void) )
 {
-	exception_t e;
-	
-	if(!__exception_handler) { return; }
+	for (int i=0;i<MAX_RESET_HANDLERS;i++)
+	{		
+		if (!__prenmi_handlers[i])
+		{
+			__prenmi_handlers[i] = cb;
+			return;
+		}
+	}
+	assertf(0, "Too many pre-NMI handlers\n");
+}
 
-	__fetch_regs(&e, EXCEPTION_TYPE_RESET, regs);
-	__exception_handler(&e);
+/** 
+ * @brief Check whether the RESET button was pressed and how long we are into
+ *        the reset process.
+ * 
+ * This function returns how many ticks have elapsed since the user has pressed
+ * the RESET button, or 0 if the user has not pressed it.
+ * 
+ * It can be used by user code to perform actions during the RESET
+ * process (see #register_reset_handler). It is also possible to simply
+ * poll this value to check at any time if the button has been pressed or not.
```

Comment:
In that case we would expect a larger than 0 value right? Being able to call this without an actual reset feels a little weird to me. Would making the initial value -1 (thus changing this to an int) make sense to detect via polling? Or maybe I'm overthinking...

Reply:
Yeah I know that -1 feels "a bit more" out of band than 0, but practically speaking 0 is out of band too (there is no way to observe a 0 value returned by this function after RESET was pressed), and 0 value is false in C which means that code is a bit more readable IMO.